src/transformers/utils/chat_template_utils.py

```diff
@@ -0,0 +1,309 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+import json
+import re
+from typing import Any, Union, get_args, get_origin, get_type_hints
+
+
+BASIC_TYPES = (int, float, str, bool, Any, type(None), ...)
+description_re = re.compile(r"^(.*?)[\n\s]*(Args:|Returns:|Raises:|\Z)", re.DOTALL)
+args_re = re.compile(r"\n\s*Args:\n\s*(.*?)[\n\s]*(Returns:|Raises:|\Z)", re.DOTALL)
+args_split_re = re.compile(
+    r"""
+(?:^|\n)  # Match the start of the args block, or a newline
+\s*(\w+):\s*  # Capture the argument name and strip spacing
+(.*?)\s*  # Capture the argument description, which can span multiple lines, and strip trailing spacing
+(?=\n\s*\w+:|\Z)  # Stop when you hit the next argument or the end of the block
+""",
+    re.DOTALL | re.VERBOSE,
+)
+returns_re = re.compile(r"\n\s*Returns:\n\s*(.*?)[\n\s]*(Raises:|\Z)", re.DOTALL)
+
+
+def _get_json_schema_type(param_type):
+    type_mapping = {
+        int: {"type": "integer"},
+        float: {"type": "number"},
+        str: {"type": "string"},
+        bool: {"type": "boolean"},
+        Any: {},
+    }
+    return type_mapping.get(param_type, {"type": "object"})
+
+
+def _parse_type_hint(hint):
+    origin = get_origin(hint)
+    args = get_args(hint)
+
+    if origin is None:
+        return _get_json_schema_type(hint)
+
+    if origin is Union:
+        # If it's a union of basic types, we can express that as a simple list in the schema
+        if all(t in BASIC_TYPES for t in args):
+            return_dict = {"type": [_get_json_schema_type(t)["type"] for t in args if t not in (type(None), ...)]}
+            if len(return_dict["type"]) == 1:
+                return_dict["type"] = return_dict["type"][0]
+        else:
+            # A union of more complex types requires us to recurse into each subtype
+            return_dict = {
+                "anyOf": [_parse_type_hint(t) for t in args if t not in (type(None), ...)],
```

Comment:
I think one if/else block can be shortcircuited if you do something like
```python
types = [_get_json_schema_type(t)["type"] for t in args if t not in (type(None), ...)]
```
then, if len(types) is one you get the simple return_dict, else you go to the recursive branch anyways?